scenario.py

```diff
@@ -189,33 +191,225 @@ def __init__(self, params, experiment_path):
+
+        # print(dict_shared)  # DEBUG
+        # print(nb_samples_needed_per_cluster)  # DEBUG
+
+        # Find if a cluster is requested more samples than it got, and if yes by which factor
+        bigger_need = 1
+        for cl in nb_samples_needed_per_cluster:
+            # print(nb_samples_needed_per_cluster)  # DEBUG
+            # print(count_per_cluster)  # DEBUG
+            bigger_need = min(bigger_need, count_per_cluster[cl] / nb_samples_needed_per_cluster[cl])
+
+        # Compute the final resize factor
+        final_resize_factor = resize_factor * bigger_need
+        # print("final_resize_factor: ", final_resize_factor)  # DEBUG
+
+        # Size correctly each partner's subset. For each partner:
+        final_nb_samples = [None] * self.partners_count
+        for p in partners_with_specific_clusters:
+            nb_samples = int(amounts_per_partner[p.id] * len(y_train) * final_resize_factor)
+            dict_specific[p.id]["final_nb_samples"] = nb_samples
+            final_nb_samples[p.id] = nb_samples
+            dict_specific[p.id]["final_nb_samples_p_cluster"] = int(
+                dict_specific[p.id]["final_nb_samples"] / p.cluster_count)
+        for p in partners_with_shared_clusters:
+            nb_samples = int(amounts_per_partner[p.id] * len(y_train) * final_resize_factor)
+            dict_shared[p.id]["final_nb_samples"] = nb_samples
+            final_nb_samples[p.id] = nb_samples
+            dict_shared[p.id]["final_nb_samples_p_cluster"] = int(
+                dict_shared[p.id]["final_nb_samples"] / p.cluster_count)
+        total_nb_samples = sum(final_nb_samples)
+        final_relative_nb_samples = [round(partner_nb_samples / total_nb_samples, 2) for partner_nb_samples in
+                                     final_nb_samples]
+        # print(dict_specific)  # DEBUG
+        # print(dict_shared)  # DEBUG
+        print(final_nb_samples)  # DEBUG
+        print(final_relative_nb_samples)  # DEBUG
+
+        # Partners receive their subsets and Partner objects are instantiated
+
+        for p in partners_with_specific_clusters:
+            list_arrays_x, list_arrays_y = [], []
+            for cl in dict_specific[p.id]["clusters_list"]:
+                list_arrays_x.append(x_train_for_cluster[cl][:dict_specific[p.id]["final_nb_samples_p_cluster"]])
+                list_arrays_y.append(y_train_for_cluster[cl][:dict_specific[p.id]["final_nb_samples_p_cluster"]])
+            p.x_train = np.concatenate(list_arrays_x)
+            p.y_train = np.concatenate(list_arrays_y)
+            p.x_test = x_test
+            p.y_test = y_test
+            # print(p_y_train)  # DEBUG
+
+        for p in partners_with_shared_clusters:
+            list_arrays_x, list_arrays_y = [], []
+            for cl in dict_shared[p.id]["clusters_list"]:
+                list_arrays_x.append(x_train_for_cluster[cl][:dict_shared[p.id]["final_nb_samples_p_cluster"]])
```

Comment:
`[:dict_specific[p.id]["final_nb_samples_p_cluster"]` looks like you always get the first part of the cluster

Reply:
Fixed